sql: decrease vectorize_row_count_threshold to 0

yuzefovich opened the pull request:
**colexec: add context to errors from builtin functions**

This commit wraps the errors that occur during builtin functions
evaluations to provide more context.

Release note: None

**sql: decrease vectorize_row_count_threshold to 0**

This commit decreases the default value for
`vectorize_row_count_threshold` setting to 0 which means that we will be
using the vectorized engine for all supported queries. We intend to
remove that setting entirely in 21.1 release, but for now we choose the
option of effectively disabling it, just in case.

The benchmarks have shown the following:
- -1.5% on KV95
- similar performance on TPCC
- -3% on movr
- -10% on miscellaneous operations (joins, aggregations) on small tables.

We think that such gap is small enough to merge this change, and we
intend to optimize the vectorized engine more before making the final
call for the default value for the 21.1 release.

Additionally, this commit collects the trace metadata on the outboxes.

Informs: #53893.

Release note (sql change): The default value for
`vectorize_row_count_threshold` setting has been decreased from 1000 to
0 meaning that from now on we will always use the vectorized engine for
all supported queries regardless of the row estimate (unless
`vectorize=off` is set).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55713)
<!-- Reviewable:end -->

yuzefovich commented:
The benchmarks.

----
3 node roachprod cluster
----
1. TPCC, 100 warehouses, 1 min ramp, 5 min duration: **-0.748%** in tpmC
  - old
```
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s     1203.2  93.6%     28.3     26.2     37.7     46.1     71.3    151.0
```
  - new
```
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
  300.0s     1194.2  92.9%     29.9     28.3     39.8     48.2     65.0    117.4
```
2. KV95, 1 min ramp, 5 min duration: **-6.45%** in throughput
  - old
```
Highest sequence written: 126430. Can be passed as --write-seq=R126430 to the next run.

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  300.0s        0        2142885         7143.0      1.0      1.0      1.8      3.0    159.4  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  300.0s        0         113060          376.9      2.5      2.4      4.1      7.3     30.4  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  300.0s        0        2255945         7519.8      1.1      1.0      2.4      3.5    159.4  
```
- new
```
Highest sequence written: 118557. Can be passed as --write-seq=R118557 to the next run.

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  300.0s        0        2004183         6680.6      1.1      1.0      2.0      3.5    151.0  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  300.0s        0         106178          353.9      2.6      2.5      4.1      6.6     23.1  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  300.0s        0        2110361         7034.5      1.1      1.0      2.5      3.9    151.0  
```
3. KV95, 5 min ramp, 30 min duration: **-8.87%** in throughput
- old
```
Highest sequence written: 843826. Can be passed as --write-seq=R843826 to the next run.

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1800.0s        0       14648469         8138.0      0.9      0.9      1.6      2.4     50.3  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1800.0s        0         771129          428.4      2.1      2.0      3.1      4.7     56.6  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
 1800.0s        0       15419598         8566.4      0.9      0.9      1.9      2.8     56.6 
```
- new
```
Highest sequence written: 774506. Can be passed as --write-seq=R774506 to the next run.

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1800.0s        0       13347899         7415.5      1.0      0.9      1.7      2.9    142.6  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
 1800.0s        0         703820          391.0      2.3      2.2      3.7      5.8     31.5  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
 1800.0s        0       14051719         7806.5      1.0      1.0      2.1      3.3    142.6 
```
----

I trust these numbers a lot more than the ones I got [here](https://github.com/cockroachdb/cockroach/pull/52453#issuecomment-670306184) because benchmarking on the laptop can have significant variance in performance. My current guess is that this difference is mostly caused by the fact that we're pooling `tableReader` objects as well as reusing some of the slices in `tableInfo` objects in the row-by-row engine. I looked briefly into how we could do the same in the vectorized engine, and especially I wanted to remove a lot of duplication between `cFetcher` and `row.fetcher` (as well as `cTableInfo` and `tableInfo`), but it's not that easy.

jordanlewis commented:
Thanks for running these benchmarks. What setup did you use? I guess we're not quite ready to merge this yet.

yuzefovich commented:
It was 3 node roachprod cluster with the default hardware options in GCE, the load was coming from node 1. Is this what you mean by "setup"?

I took a look at profiles, and nothing really stood out, and the pooling behavior is the main difference I can see right now. I'll take a stab tomorrow at cleaning up the fetchers a bit.

jordanlewis commented:
Yes, I meant hardware setup. Thanks!

awoods187 commented:
On tpcc we typically use a fourth node for the workload to simulate an
application being separate from the database resources. May be worth
adjusting the testing setup in the future.

On Wed, Oct 21, 2020 at 11:30 PM Jordan Lewis <notifications@github.com>
wrote:

> Yes, I meant hardware setup. Thanks!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/55713#issuecomment-714198363>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFJ7F75CGTQ32LCAJ6BVEMLSL6ROZANCNFSM4SWWJKFQ>
> .
>
-- 
Andy Woods
Senior Product Manager, Cockroach Labs
andy@cockroachlabs.com

yuzefovich commented:
One pretty obvious thing that I missed that is likely to explain the majority of the difference is that without `vectorize_row_count_threshold` we are now always running `SupportsVectorized` check (previously, we would do so only if the threshold is met). That check especially shows up on KV-like workload because we're essentially instantiating the whole flow twice with only the second time running it, and in particular that means that we're creating a `ColBatchScan` and initializing `cFetcher` twice too - `cFetcher.Init` is pretty heavy.

So I think apart from the pooling of objects we'll need to do something about this check. I feel like the most viable option is to make it a lot more lightweight: instead of creating the whole fake flow with all of the components, we could probably just inspect the processor specs and see whether there are any that we refuse to wrap, and if we either support "natively" or can wrap all of the cores, then we'll assume that we can successfully setup the whole flow. I think when `SupportsVectorized` check was introduced, we had a lot more edge cases that needed to be handled, so we went with the fake flow setup, but now that might be unnecessary anymore. It does feel a bit scary to remove that check though.

yuzefovich commented:
Hm, to my surprise a prototype with some pooling and a more lightweight support check (#55883) shows big drop in performance between `vectorize_row_count_threshold=1000` and `vectorize_row_count_threshold=0`. I guess we'll have to abandon this PR for now :/

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55713)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/cli/cli_test.go, line 531 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55713#-MKKSgrFDpGYk5m_BiZV:-MKKSgrFDpGYk5m_BiZW:b-mf01x3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d38f53e73fc5c4187b50c441ce49a81aae28cf16/pkg/cli/cli_test.go#L531)):*
> ```Go
> 	c.RunWithArgs([]string{`sql`, `--set`, `unknownoption`, `-e`, `select 123 as "123"`})
> 	// Check that partial results + error get reported together.
> 	c.RunWithArgs([]string{`sql`, `-e`, `select 1/(@1-2) from generate_series(1,3)`})
> ```

Out of curiosity, why did you need to change this?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55713)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/cli/cli_test.go, line 531 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55713#-MKKSgrFDpGYk5m_BiZV:-MKKxxMeDrwpU3dE86I4:b4fa1wk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d38f53e73fc5c4187b50c441ce49a81aae28cf16/pkg/cli/cli_test.go#L531)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Out of curiosity, why did you need to change this?
</blockquote></details>

Currently the test will fail if run via the vectorized engine because a division by zero error will occur sooner (it occurs on the third row, but we have batches of growing capacity starting at 1, so the batch sizes will be 1, 2, 4, ..., 1024, 1024, ..., and the error occurs when processing the second batch at which point only 1 row has been returned to the client whereas the test expects 2 rows).


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55713)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/cli/cli_test.go, line 531 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55713#-MKKSgrFDpGYk5m_BiZV:-MKL05Ke2QxwRZOV_X75:bxxkz7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d38f53e73fc5c4187b50c441ce49a81aae28cf16/pkg/cli/cli_test.go#L531)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Currently the test will fail if run via the vectorized engine because a division by zero error will occur sooner (it occurs on the third row, but we have batches of growing capacity starting at 1, so the batch sizes will be 1, 2, 4, ..., 1024, 1024, ..., and the error occurs when processing the second batch at which point only 1 row has been returned to the client whereas the test expects 2 rows).
</blockquote></details>

I think it's worth explaining this in a comment here.


<!-- Sent from Reviewable.io -->

asubiotto commented:
Let's discuss this and how to move forward during the next meeting.

yuzefovich commented:
A newer prototype (that is almost production ready) shows an [improvement](https://github.com/cockroachdb/cockroach/pull/55948#issuecomment-716083162) - the performance hit on KV95 if we remove `SupportsVectorized` check in favor of a more light-weight one decreases from 6-8% to 3-4%.

yuzefovich committed:
sql: decrease vectorize_row_count_threshold to 0

This commit decreases the default value for
`vectorize_row_count_threshold` setting to 0 which means that we will be
using the vectorized engine for all supported queries. We intend to
remove that setting entirely in 21.1 release, but for now we choose the
option of effectively disabling it, just in case.

The benchmarks have shown the following:
- -1.5% on KV95
- similar performance on TPCC
- -3% on movr
- -10% on miscellaneous operations (joins, aggregations) on small tables.

We think that such gap is small enough to merge this change, and we
intend to optimize the vectorized engine more before making the final
call for the default value for the 21.1 release.

Additionally, this commit collects the trace metadata on the outboxes.

Release note (sql change): The default value for
`vectorize_row_count_threshold` setting has been decreased from 1000 to
0 meaning that from now on we will always use the vectorized engine for
all supported queries regardless of the row estimate (unless
`vectorize=off` is set).

yuzefovich reviewed (commented):
Alright, I think this work can move forward, and the PR is RFAL.

I'm currently rerunning KV95 benchmarks, but recent results [here](https://github.com/cockroachdb/cockroach/pull/55948#issuecomment-716083162) have shown that the performance hit is now reduced from 6-8% to 3-4%, and we deemed it acceptable given our intention to optimize the read path more during this release.

I also want to run some other benchmarks making sure that we don't regress on other types of queries (TPCC workload, some light lookup joins, hash joins, aggregations).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55713#-:-MKl1EGNAm_oSdEJ9VKt:b-mqf4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/cli/cli_test.go, line 531 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/55713#-MKKSgrFDpGYk5m_BiZV:-MKl0ywSBxmrbj1Zuk49:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d38f53e73fc5c4187b50c441ce49a81aae28cf16/pkg/cli/cli_test.go#L531)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think it's worth explaining this in a comment here.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

awoods187 commented:
I'm not sure a 3-4% hit is acceptable for our most important workloads. Have we engaged with KV and other stakeholders on this change? cc @nvanbenschoten

jordanlewis commented:
We've planned to merge this early in the cycle, communicate out to the team, and evaluate by the end of the cycle whether or not we can keep this setting as is. Certainly, we are not interested in merging something like this without approval from stakeholders.

yuzefovich commented:
The latest KV95 numbers.
1. 2 min ramp, 10 min duration, **-3.25%**
- `vectorize_row_count_threshold=1000`:
```
 600.0s        0        5279133         8798.6      0.9      0.9      1.9      2.8     29.4 
```
- `vectorize_row_count_threshold=0`:
```
  600.0s        0        5107456         8512.4      0.9      0.9      1.9      2.9     39.8 
```
2. 5 min ramp, 30 min duration, **-1.69%**
- `vectorize_row_count_threshold=1000`:
```
 1800.0s        0       16038905         8910.5      0.9      0.9      1.8      2.6    121.6
```
- `vectorize_row_count_threshold=0`:
```
 1800.0s        0       15767558         8759.8      0.9      0.9      1.8      2.6    151.0 
```

I added a bit more pooling of objects since the previous run, and it seemed to have decreased the performance hit.

As Jordan said, we intend on changing the default value without actually removing the setting so that we can fallback to the old status quo if we need to. However, I'm quite confident that with a bit more optimization of the critical read only path in the vectorized engine we should be able to get to parity. The reason for changing the default value this early into the release is that some of the optimizations (namely the removal of `SupportsVectorized` check that is part of this PR) might be a bit destabilizing in some edge cases, and we hope to flush those out during the "baking period." One example is a wrong assumption that I put in place that was exposed by the CI on this PR and is now fixed in the third commit.

nvanbenschoten commented:
This is exciting! Thanks for running all of these benchmarks @yuzefovich.

> I'm not sure a 3-4% hit is acceptable for our most important workloads. Have we engaged with KV and other stakeholders on this change? cc @nvanbenschoten

I'm a little unsure about what we gain from the last commit. We won't be able to remove the `vectorize_row_count_threshold` session variable (https://github.com/cockroachdb/cockroach/issues/53893) until we close the gap entirely, or at least lower the regression to well below 1%, so what's the benefit of making this change now instead of after we've closed the gap?

I don't really have a strong opinion here, other than that it seems like more work down the line to remember to continually check whether the gap has been closed by flipping this setting back to 1000. We won't be able to rely on roachperf for this, because that will pick up other changes over the course of this release.

jordanlewis commented:
I think making the change early will make the worse behavior show up more prominently in profiles, and nudge us to improve the setup path. Getting rid of this setting is an important goal, since it further paves our way to removing the non-vectorized engine altogether, which will reduce code duplication, tech debt, and potential for confusing user-facing behavior.

We'd also like to make this change early to catch more issues, earlier in the cycle. Setting this threshold to 0 will make it more likely we'll discover any remaining bugs we haven't discovered via our other testing.

Here's a hypothetical for you, Nathan. Let's say we gained 5% performance elsewhere in the system, and lost 4% net on flow setup, leaving us at 1% gain in 21.1. How would we decide whether to leave this setting on or off? Do we pursue performance at all costs, or do we leave room for projects that help CockroachDB in other ways?

I'm not sure about my answer to this hypothetical in general, but I think in this particular scenario I'd be fine accepting the 1% gain, instead of the slightly larger gain, given the other benefits.

nvanbenschoten commented:
That depends entirely on what's improving and what we're gaining in exchange for the lost efficiency. Since I don't know the full scope of the "other benefits" of this change, I can't make any determination there. Eliminating duplicate code is very compelling.

> I think making the change early will make the worse behavior show up more prominently in profiles, and nudge us to improve the setup path.
> We'd also like to make this change early to catch more issues, earlier in the cycle.

I buy both of those arguments.

> How would we decide whether to leave this setting on or off? Do we pursue performance at all costs, or do we leave room for projects that help CockroachDB in other ways?

On an important enough workload (which I'd argue KV is, given that we're primarily an OLTP database and KV is a proxy for most other OLTP workloads), I don't think there's a huge difference between a 1% gain that could have been 5% or a 4% regression. Of course, we couldn't release with the regression and could with the limited gain and so there's a timing question. But either way, we're now 4% less efficient and 4% more expensive to run than we know we could be by not making the change, so the justification for such a regression needs to be just as strong, regardless of whether it was offset or not.

yuzefovich commented:
Thanks everyone for the input. We've decided to not merge this change as is (with the performance hit), and, instead, I'll spend more time on this PR optimizing the vectorized flow setup / read path in the next week or two planning at least to get to the parity with the row-based flow on all of the important workloads.

yuzefovich commented:
Here is what I got after rerunning the benchmarks with all the latest changes:
- KV95 (full output [here](https://gist.github.com/yuzefovich/d99730c2f70559307dc81a87adb611d7)):
  - 1 min ramp, 5 min duration
    - GCE: **-1.11%**
    - AWS: **-6.49%**
  - 5 min ramp, 30 min duration
    - GCE: **-2.32%**
    - AWS: **-1.65%**
  - 10 min ramp, 60 min duration
    - GCE: **-3.19%**
    - AWS: **+0.33%**
- TPCC 100 (full output [here](https://gist.github.com/yuzefovich/55a29e043d77c419fb3af96c961385a4)):
  - 3 min ramp, 15 min duration
    - GCE: **+0.2% efc** and **+4.8% in P50**
    - AWS: **+0.0% efc** and **+10.6% in P50**
  - 10 min ramp, 60 min duration
    - GCE: **+0.1% efc** and **-1.8% in P50**
    - AWS: **+0.1% efc** and **-4.5% in P50**

(Note that I'd disregard AWS 5 min number since it is likely to be too noisy - in my runs I hadn't seen this big of a difference since the change to the SupportsVectorized check.)

I had been running KV95 benchmark many times in the last couple of weeks, and usually the longer the duration, the smaller the performance hit. I would estimate that, if we perform many runs with duration of 30 minutes or more, the performance hit is on the order of 1.5% on average.

This was the first time I looked at TPCC numbers, and they are pretty good.

I merged all the optimizations that I could find in the vectorized flow setup path, so we should be making the decision whether  to decrease the threshold to 0 given these numbers. I'm in favor of moving forward with the change given KV95 workload is expected to be the worst case for the vectorized engine and the performance hit. If we come to an agreement that 1.5% or so hit there is acceptable, I'll also confirm the performance difference on other queries (some joins and aggregations on small tables), but I don't expect to have a performance hit there.

asubiotto commented:
Thanks for running these benchmarks. I think that we should go ahead with this change. What we lose in a slight hit to performance we gain in moving forward on the path to removing the row execution engine and having to maintain two execution engines, as well as the better stability the vectorized engine gives us in terms of memory management.

yuzefovich commented:
Hm, I ran some queries with different operations (merge and hash joins, ordered and hash aggregations) on small datasets (10, 100, 1k rows) on 3 node cluster, and the vectorized engine's performance catches up only at 1k, on smaller sets there is 1.3-2x hit in the throughput. Full results are [here](https://gist.github.com/yuzefovich/3156a40bacbe64a1d2978b241251edf1).

The results are somewhat surprising to me (I haven't looked at the profiles though, and I thought the dynamic batch size work would take of most of the difference, yet it is likely that the difference is explained by larger internal allocations). I want to spend more time looking into this.

yuzefovich commented:
Quick update here is that actually the results I obtained on miscellaneous queries are expected - in all of the tested operations we still have some large fixed-size allocations that weren't resolved as part of the work to introduce the dynamic batch size. I'm currently working on addressing those and making the vectorized engine fully dynamic with regards to the batches it operates on which - hopefully - will bring us to parity on those queries on small datasets.

yuzefovich commented:
I reran the miscellaneous queries using all of the latest  changes to make the vectorized engine more dynamic (including the not-yet-merged constants tuning PR) and updated the [results](https://gist.github.com/yuzefovich/3156a40bacbe64a1d2978b241251edf1). The vectorized engine is now on the order of 10% slower on sets of 10 and 100 rows and is slightly faster at 1k rows. These results seem good enough for me.

I also ran a couple of quick KV95 benchmarks, and I got:
- 1 min ramp, 5 min duration
  - GCE: **-3.64%**
  - AWS: **+0.13%**
- 3 min ramp, 15 min duration
  - GCE: **-0.49%**
  - AWS: **-0.37%**

I think I'm finally satisfied with all of the numbers and would like to proceed with this change. Thoughts?

yuzefovich commented:
More KV95 numbers:
- 1 min ramp, 5 min duration
  - GCE: **+3.68%**
  - AWS: **-3.84%**
- 2 min ramp, 10 min duration
  - GCE: **-2.84%**
  - AWS: **-1.10%**
- 3 min ramp, 15 min duration
  - GCE: **+3.58%**
  - AWS: **-3.70%**
- 5 min ramp, 30 min duration
  - GCE: **-2.51%**
  - AWS: **-0.66%**
- 10 min ramp, 60 min duration
  - GCE: **+1.37%**
  - AWS: **-1.87%**

jordanlewis commented:
LGTM! Ship it!

jordanlewis commented:
Err, I guess we decided to also see TPCC numbers, right? I'm getting eager.

yuzefovich commented:
Yes, I'll be rerunning TPCC today as well as MovR to see what's up, and if those look good, we'll merge tomorrow I think.

yuzefovich commented:
TPCC100 numbers again show no significant difference:
- 1 min ramp, 5 min duration
  - GCE:
    - old: `300.0s     1193.2  92.8%     24.1     23.1     30.4     37.7     54.5     92.3`
    - new: `300.0s     1194.8  92.9%     23.9     23.1     30.4     37.7     52.4     96.5`
  - AWS:
    - old: `300.0s     1204.8  93.7%     16.4     15.7     19.9     26.2     39.8    104.9`
    - new: `300.0s     1198.6  93.2%     17.3     16.3     22.0     27.3     41.9    226.5`
- 3 min ramp, 15 min duration
  - GCE:
    - old: `900.0s     1233.3  95.9%     21.9     22.0     26.2     30.4     46.1    121.6`
    - new: `900.0s     1228.7  95.5%     21.9     21.0     27.3     31.5     48.2     88.1`
  - AWS:
    - old: `900.0s     1237.5  96.2%     15.9     15.2     18.9     22.0     35.7    243.3`
    - new: `900.0s     1235.9  96.1%     15.1     14.7     17.8     21.0     33.6    100.7`
- 10 min ramp, 60 min duration
  - GCE:
    - old: `3600.0s     1251.7  97.3%     23.1     23.1     27.3     31.5     48.2    130.0`
    - new: `3600.0s     1252.0  97.4%     22.5     22.0     26.2     30.4     48.2    151.0`
  - AWS:
    - old: `3600.0s     1255.1  97.6%     15.3     14.7     17.8     19.9     33.6    436.2`
    - new: `3600.0s     1253.5  97.5%     15.7     15.2     17.8     19.9     33.6    226.5`

The only interesting detail is that during 60 min run with *old* config on AWS apparently one node died and "new order" transactions were returning an error. I didn't notice it before wiping the cluster, so I'm not sure what happened.

----
And here are the numbers of `movr` workload (with these parameters `--num-histories 10000 --num-promo-codes 10000 --num-rides 5000     --num-users 500 --num-vehicles 150`):
**TODO: this part is incomplete**
- 1 min ramp, 5 min duration
  - GCE:
    - old: `300.0s        0          88306          294.4      3.4      1.2     12.1     44.0     96.5`
    - new: `300.0s        0          86214          287.4      3.5      1.4     11.5     41.9     60.8`
    - **-2.37%** in throughput
  - AWS:
    - old: ``
    - new: ``
    - **** in throughput
- 3 min ramp, 15 min duration
  - GCE:
    - old: `900.0s        0         260376          289.3      3.5      1.4     11.5     41.9     67.1`
    - new: `900.0s        0         254317          282.6      3.5      1.4     13.1     41.9     60.8`
    - **-2.33%** in throughput
  - AWS:
    - old: ``
    - new: ``
    - **** in throughput
- 10 min ramp, 60 min duration
  - GCE:
    - old: `3600.0s        0        1077971          299.4      3.3      1.6     10.0     37.7    209.7`
    - new: ``
    - **** in throughput
  - AWS:
    - old: `3600.0s        0        1488654          413.5      2.4      1.4      6.8     22.0    436.2`
    - new: ``
    - **** in throughput

yuzefovich commented:
Updated all the benchmark numbers and the commit message. PTAL, @asubiotto @jordanlewis

yuzefovich committed:
colexec: add context to errors from builtin functions

This commit wraps the errors that occur during builtin functions
evaluations to provide more context.

Release note: None

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55713#-:-MNc5h6c-r3GnDHUm56p:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+

yuzefovich commented:
`test` target timed out, `testrace` target is a flake that will be fixed shortly. I'll amend the commit and force-push in order to run the CI again before merging to see whether the time out is related or not.

yuzefovich commented:
Ok, `test` target passed, `testrace` target should be fixed.

bors r+
